### PR TITLE
CDAP-2761: instrumentation of Cube and metrics code

### PR DESCRIPTION
**Pull request by abaranau**
Tested in local, distributed modes.

Couple notes:

* You'll find in watchdog a lot of passing thru of MetricsContext or MetricsCollector. As per discussion with Trex we decided to use that approach first. The alternative would be using approach similar to logging: pass context thru inheritable threadlocal. Current metrics stuff differs from logging (at least) by changing context, including in same thread. In future, we will likely do same for logging (as discussed like 2 years ago:)) - in any case we'll likely use unified approach. We'll see how - as we have more cases.

* Some metrics processing metrics are written directly to store, some are going thru normal aggregation (Kafka). For now, only "operating" metrics, e.g. metrics processing latency and # of processed metrics are emitted directly: we want to see these right away, without any delay. Others are collected in re-usable components and it is natural easiest way to emit those in a standard way. That means in idle system couple metrics (less than 10) are always aggregated and processed. Which is fine. But we may improve in future.

* Added metrics names may seem confusing or need to go into code to understand and interpret. We may improve naming in future. Note that those are "debug" level metrics - you only need those during heavy optimization and debugging.. So for now - it is ok.

* Cube dataset instrumentation (separate from metrics code - txnl) to be tested separatelly.

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2123-1 GREEN

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java [hunk 1]
```diff
@@ -76,15 +79,22 @@ public FactTable get(int resolution, int ignoredRollTime) {
...
       }
     };
+
     this.cube = Suppliers.memoize(new Supplier<Cube>() {
       @Override
       public Cube get() {
-        // 1 sec, 1 min, 1 hour and "all time totals"
-        return new DefaultCube(resolutions, factTableSupplier, createAggregations());
+        DefaultCube cube = new DefaultCube(resolutions, factTableSupplier, createAggregations());
+        cube.setMetricsCollector(metricsContext);
```
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: We should add capabilities to turn off the instrumentation.

**Comment by sreevatsanraman** [comment]
The changes look good, however we should have capabilities to turn on/off instrumentation. By default should be off.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java [hunk 2]
```diff
@@ -82,12 +85,14 @@ public void add(CubeFact fact) {
...
   public void add(Collection<? extends CubeFact> facts) {
     List<Fact> toWrite = Lists.newArrayList();
+    int dimValuesCount = 0;
     for (CubeFact fact : facts) {
       for (Aggregation agg : aggregations.values()) {
         if (agg.accept(fact)) {
           List<DimensionValue> dimensionValues = Lists.newArrayList();
           for (String dimensionName : agg.getDimensionNames()) {
             dimensionValues.add(new DimensionValue(dimensionName, fact.getDimensionValues().get(dimensionName)));
+            dimValuesCount++;
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: can do, dimValuesCount += dimensionValues.size() after this for loop, i am not sure if this metric will be helpful though.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java [hunk 3]
```diff
@@ -173,7 +193,14 @@ SELECT count('read.ops')                                           << measure na
     FactTable table = resolutionToFactTable.get(query.getResolution());
     FactScanner scanner = table.scan(scan);
     Table<Map<String, String>, String, Map<Long, Long>> resultMap = getTimeSeries(query, scanner);
-    return convertToQueryResult(query, resultMap);
+
+    incrementMetric("cube.query.request.success.count", 1);
+    incrementMetric("cube.query.result.map.size", resultMap.size());
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: table.size ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java [hunk 4]
```diff
@@ -154,11 +169,16 @@ SELECT count('read.ops')                                           << measure na
...
       agg = findAggregation(query);
       if (agg == null) {
+        incrementMetric("cube.query.request.failure.count", 1);
         throw new IllegalArgumentException("There's no data aggregated for specified dimensions " +
                                              "to satisfy the query: " + query.toString());
       }
     }
 
+    // tell how many queries end up querying specific pre-aggregated views and resolutions
+    incrementMetric("cube.query.agg." + agg + ".count", 1);
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: how does this metric name look like ? , i don't see toString in DefaultAggregation

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/timeseries/FactTable.java [hunk 5]
```diff
@@ -127,6 +135,10 @@ public void add(List<Fact> facts) {
     // todo: replace with single call, to be able to optimize rpcs in underlying table
     timeSeriesTable.put(convertedGaugesTable);
     timeSeriesTable.increment(convertedIncrementsTable);
+    if (metrics != null) {
+      metrics.increment("factTable.put.count", convertedGaugesTable.size());
+      metrics.increment("factTable.increment.count", convertedIncrementsTable.size());
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/timeseries/FactTable.java]: would it be useful to add the resolution of the factTable in the metric name ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java [hunk 6]
```diff
@@ -120,6 +130,9 @@ public Cube get() {
     aggs.put("flow.queue", new DefaultAggregation(
       ImmutableList.of(Constants.Metrics.Tag.NAMESPACE, Constants.Metrics.Tag.APP,
                        Constants.Metrics.Tag.FLOW, Constants.Metrics.Tag.CONSUMER,
+                       Constants.Metrics.Tag.PRODUCER, Constants.Metrics.Tag.FLOWLET_QUEUE),
+      ImmutableList.of(Constants.Metrics.Tag.NAMESPACE, Constants.Metrics.Tag.APP,
+                       Constants.Metrics.Tag.FLOW, Constants.Metrics.Tag.CONSUMER,
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: unrelated change ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java [hunk 7]
```diff
@@ -44,21 +46,22 @@
...
   public MetricsMessageCallback(DatumReader<MetricValues> recordReader,
                                 Schema recordSchema,
-                                MetricStore metricStore) {
+                                MetricStore metricStore,
+                                @Nullable
+                                MetricsContext metricsContext) {
     this.recordReader = recordReader;
     this.recordSchema = recordSchema;
     this.metricStore = metricStore;
+    this.metricsContext = metricsContext == null ? Collections.<String, String>emptyMap() : metricsContext.getTags();
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: what does an empty map mean, they don't have any context? , then it becomes inefficient to search/query them right ?

**Comment by abaranau** [comment]
> The changes look good, however we should have capabilities to turn on/off instrumentation. By default should be off.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java (hunk 2 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: yeah, not sure :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java (hunk 4 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: oh shoot. Nice catch. It should be agg name :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java (hunk 3 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java]: result.size

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 7 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: yes, there could be empt context. It is accepting null for conveninece

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: it may be not super efficient, but fine, if that is what desired

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java (hunk 1 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: why?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java (hunk 6 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java]: well, while testing and debugging - found that we do write too many data than needed - i.e. that filter was absent. It needs to be there. Fine within this PR:)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/timeseries/FactTable.java (hunk 5 above)
Comment (abaranau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/timeseries/FactTable.java]: yes it will be useful!

**Comment by abaranau** [comment]
@sreevatsanraman @gsps1 please take another look

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2123-2

**Comment by sreevatsanraman** [comment]
> Why do we ever want to turn it off? Esp. by default. Why?
Discussed with Sasha, since we are going to be emitting fewer metrics (~10) compared to the rest of the system/app metrics, this is no longer a concern.

**Comment by abaranau** [comment]
@gsps1 please take a look

**Comment by gsps1** [comment]
LGTM

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2123-5 GREEN
